diff --git a/kernel/feature/sucompat.c b/kernel/feature/sucompat.c
index 6dae885..05c477f 100644
--- a/kernel/feature/sucompat.c
+++ b/kernel/feature/sucompat.c
@@ -5,7 +5,13 @@
 #include <linux/preempt.h>
 #include <linux/printk.h>
 #include <linux/mm.h>
+#if defined(__has_include)
+#if __has_include(<linux/pgtable.h>)
 #include <linux/pgtable.h>
+#endif
+#else
+#include <linux/pgtable.h>
+#endif
 #include <linux/uaccess.h>
 #include <asm/current.h>
 #include <linux/cred.h>
