diff --git a/kernel/infra/seccomp_cache.c b/kernel/infra/seccomp_cache.c
index 6511066..74fea37 100644
--- a/kernel/infra/seccomp_cache.c
+++ b/kernel/infra/seccomp_cache.c
@@ -1,4 +1,5 @@
 #include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
 #include <linux/fs.h>
 #include <linux/nsproxy.h>
 #include <linux/sched/task.h>
@@ -6,7 +7,7 @@
 #include <linux/filter.h>
 #include <linux/seccomp.h>
 #include "klog.h" // IWYU pragma: keep
-#include "infra/seccomp_cache.h"
+#include "seccomp_cache.h"
 
 struct action_cache {
 	DECLARE_BITMAP(allow_native, SECCOMP_ARCH_NATIVE_NR);
@@ -62,4 +63,5 @@ void ksu_seccomp_allow_cache(struct seccomp_filter *filter, int nr)
         set_bit(nr, filter->cache.allow_compat);
     }
 #endif
-}
\ No newline at end of file
+}
+#endif // #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
\ No newline at end of file
diff --git a/kernel/infra/seccomp_cache.h b/kernel/infra/seccomp_cache.h
index 270381d..5900f28 100644
--- a/kernel/infra/seccomp_cache.h
+++ b/kernel/infra/seccomp_cache.h
@@ -1,10 +1,12 @@
-#ifndef __KSU_H_KERNEL_COMPAT
-#define __KSU_H_KERNEL_COMPAT
+#ifndef __KSU_H_SECCOMP_CACHE
+#define __KSU_H_SECCOMP_CACHE
 
 #include <linux/fs.h>
 #include <linux/version.h>
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
 extern void ksu_seccomp_clear_cache(struct seccomp_filter *filter, int nr);
 extern void ksu_seccomp_allow_cache(struct seccomp_filter *filter, int nr);
+#endif // #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
 
 #endif
